core/src/main/java/com/linecorp/armeria/common/FlagsProvider.java

```diff
@@ -0,0 +1,922 @@
+     * trace.
+     *
+     * <p>The default value of this flag is {@value DefaultFlagsProvider#VERBOSE_EXCEPTION_SAMPLER},
+     * which retains the stack trace of the exceptions at the maximum rate of 10 exceptions/sec.
+     * Specify the {@code -Dcom.linecorp.armeria.verboseExceptions=<specification>} JVM option to override
+     * the default. See {@link Sampler#of(String)} for the specification string format.</p>
+     */
+    @Nullable
+    default Sampler<Class<? extends Throwable>> verboseExceptionSampler() {
+        return null;
+    }
+
+    /**
+     * Returns whether to log the socket exceptions which are mostly harmless. If enabled, the following
+     * exceptions will be logged:
+     * <ul>
+     *   <li>{@link ClosedChannelException}</li>
+     *   <li>{@link ClosedSessionException}</li>
+     *   <li>{@link IOException} - 'Connection reset/closed/aborted by peer'</li>
+     *   <li>'Broken pipe'</li>
+     *   <li>{@link Http2Exception} - 'Stream closed'</li>
+     *   <li>{@link SSLException} - 'SSLEngine closed already'</li>
+     * </ul>
+     *
+     * <p>It is recommended to keep this flag disabled, because it increases the amount of log messages for
+     * the errors you usually do not have control over, e.g. unexpected socket disconnection due to network
+     * or remote peer issues.</p>
+     *
+     * <p>This flag is disabled by default.
+     * Specify the {@code -Dcom.linecorp.armeria.verboseSocketExceptions=true} JVM option to enable it.</p>
+     *
+     * @see Exceptions#isExpected(Throwable)
+     */
+    @Nullable
+    default Boolean verboseSocketExceptions() {
+        return null;
+    }
+
+    /**
+     * Returns whether the verbose response mode is enabled. When enabled, the server responses will contain
+     * the exception type and its full stack trace, which may be useful for debugging while potentially
+     * insecure. When disabled, the server responses will not expose such server-side details to the client.
+     *
+     * <p>This flag is disabled by default. Specify the {@code -Dcom.linecorp.armeria.verboseResponses=true}
+     * JVM option or use {@link ServerBuilder#verboseResponses(boolean)} to enable it.
+     */
+    @Nullable
+    default Boolean verboseResponses() {
+        return null;
+    }
+
+    /**
+     * Returns the fully qualified class name of {@link RequestContextStorageProvider} that is used to choose
+     * when multiple {@link RequestContextStorageProvider}s exist.
+     *
+     * <p>The default value of this flag is {@code null}, which means only one
+     * {@link RequestContextStorageProvider} must be found via Java SPI. If there are more than one,
+     * you must specify the {@code -Dcom.linecorp.armeria.requestContextStorageProvider=<FQCN>} JVM option to
+     * choose the {@link RequestContextStorageProvider}.
+     */
+    @Nullable
+    default String requestContextStorageProvider() {
+        return null;
```

Comment:
> Make DefaultFlagsProvider perform the SPI lookup and instantiation for RequestContextStorageProvider lazily (only once on demand).

to retain the previous(before this pull request) behavior and align with the new logic. I adjusted your suggestion a bit.
- Move logic from `RequestContextUtil (static block)` to `SystemPropertyFlagsProvider` perform the SPI lookup, filtering and instantiation lazily for `RequestContextStorageProvider`.
  - if no requestContextStorage providers are found, then return null to defer to the next flags provider. 
  - if none of the requestContextStorage providers matches with JVM option, then return null to defer to the next flags provider.
- At `DefaultFlagsProvider` performs SPI lookup again (reuse from the previous lookup) and throws an error if more than one provider is found, if not then use `RequestContextStorage.threadLocal()`.

I add SPI lookup logic at `DefaultFlagsProvider` because technically there is a case that requestContextStorage providers are provided but then users specify the wrong FQCN. Then we need to defer to other flags providers. In case no flags provider is found then `DefaultFlagsProvider` needs to check again if there is any requestContextStorage provider. If yes then an error is thrown at `DefaultFlagsProvider` because the system cannot choose requestContextStorage provider, else `RequestContextStorage.threadLocal()` can be used as the default. This is an edge case that I'm not sure whether I should consider or ignore.

I already push commits to pull requests. If there's something that I missed or adjustment is needed, please let me know. Thank you.